### ISSUE TITLE
[no-undef] should accept `typeof` as proof a variable exists

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#6209]
In the following code, it's statically verifiable through code path analysis that `x` must be defined in the body of the `if` statement below.

```js
if (typeof x !== "undefined") {
    doSomethingWith(x)
}
```

Most commonly, I need this conditional thing when targeting ES6 and ES5 globals, or targeting both the browser and Node. Things like this are very common:

```js
// In my tests
if (typeof Symbol === "function") {
    expect(Array.isArray(Symbol())).to.be.false
}

// In a module, if I need a map, but I have a backup class and don't need a
// fully featured and correct polyfill to do the job (albeit slower).
var map = typeof Map !== "undefined" ? new Map() : new PolyMap()
```

Could `no-undef` follow code paths and consider variables with `typeof` guards to be defined as long as the comparison implies they cannot be `undefined`? It's not fun having to suffix every line like this with `// eslint-ignore-line no-undef`, especially with the frequency I need them with several ES5 libraries.

**Comment by isiahmeadows** [comment]
To clarify, all statements in both code blocks above fail in the latest `npm` release of ESLint.

**Comment by scriptdaemon** [comment]
This is what `/*global ...*/` is for.

Ex:

```
/*global Map Symbol*/

// In my tests
if (typeof Symbol === "function") {
    expect(Array.isArray(Symbol())).to.be.false
}

// In a module, if I need a map, but I have a backup class and don't need a
// fully featured and correct polyfill to do the job (albeit slower).
var map = typeof Map !== "undefined" ? new Map() : new PolyMap()
```

**Comment by Slayer95** [comment]
@scriptdaemon, the issue intends something like

** Correct code **

```
/*eslint no-undef: ["error", {"allowImplicitGlobals":true}] */

if (typeof OptionalGlobal !== "undefined") {
    OptionalGlobal.compatUninstall();
}
```

** Incorrect code **

```
/*eslint no-undef: ["error", {"allowImplicitGlobals":true}] */

if (typeof OptionalGlobal !== "undefined") {
    OptionalGlobal.compatUninstall();
}

var result = OptionalGlobal.someMethod();
```

**Comment by scriptdaemon** [comment]
I think I see now. This is what `/*global ...*/` comments are for.

```js
/*global Map Symbol*/

// In my tests
if (typeof Symbol === "function") {
    expect(Array.isArray(Symbol())).to.be.false
}

// In a module, if I need a map, but I have a backup class and don't need a
// fully featured and correct polyfill to do the job (albeit slower).
var map = typeof Map !== "undefined" ? new Map() : new PolyMap()
```

**Comment by isiahmeadows** [comment]
I know this already, and I've done similar already in a few cases. It's still just another band-aid workaround that doesn't fix much.

I'm also looking into cases like if you target both the browser and Nashorn, and you probably don't want all those Nashorn globals glossing over potentially undefined variables in browser-specific code within a file. There's so little intersection beyond the required globals that it can potentially mask bugs.

The reason I'm requesting code path analysis to be applied is because it requires less special casing to make the linter happy, and it's a very valid deduction.

-----

*(I forgot to mention `PolyMap` would be something that's already defined within the module.)*

**Comment by ilyavolodin** [comment]
What should happen with the following code:
```js
if (typeof Symbol === "function") {
    createSymbol();
}

function createSymbol() {
    Symbol.search(.....);
}
```

While we could make some assumptions based on the code with static analysis, I'm not really sure it's worth it, in this particular case. It would be very complex code with a lot of edge cases, and could easily be replaced by just declaring a global variable.

**Comment by Slayer95** [comment]
While I would prefer myself for that use case __to warn__, I understand the impossibility of asserting that the function is always called from safe places, specially when it's exported.

So I would be happy instead to help the linter with an ``/* eslint disable no-undef */`` on top of any such functions. Or better yet, ``/* eslint no-undef ["error", {"except": ["Symbol"]}] */`` (new feature: ``except`` option, assuming that it can be setted/unsetted granulary).

However, I'd say warning in most other -simpler- cases is the very job of a linter.

**Comment by scriptdaemon** [comment]
@Slayer95 The problem with an `except` option is that it provides the exact same functionality as using `/*global ...*/`, which is redundant and is extra work for no benefit.

**Comment by Slayer95** [comment]
No, because globals cannot be unset.

**Comment by ilyavolodin** [comment]
`/* eslint */` directives are hoisted on a file bases. So you can't just set them around a function (or rather, you can, but they will apply to the whole file).

**Comment by nzakas** [comment]
This is crossing over into evaluating code, not just staticly analyzing it, and that's a lot more complicated than just code path analysis. We don't have any capacity for conditionally defining variables as we evaluate code, which would also be a prerequisite. 

👎

**Comment by Slayer95** [comment]
I got a [proof of concept working (@MIT License)](http://hastebin.com/pexapeyuku.coffee).
But I am sure it can be improved upon with ESLint's integrated path analysis tools.

**Comment by isiahmeadows** [comment]
@ilyavolodin 

In your code sample here:

```js
if (typeof Symbol === "function") {
    createSymbol();
}

function createSymbol() {
    Symbol.search(.....);
}
```

That should most definitely warn, because that function can in theory be called from outside the code block, and as @nzakas pointed out, that would literally require code execution to find out whether it's actually defined or not. I'm strictly requesting that within a single code branch, it is seen as defined, and not checked. Had it been defined like this, it should be okay in strict mode ES2015, because the function is completely localized to the block:

```js
if (typeof Symbol === "function") {
    function createSymbol() {
        Symbol.search(.....);
    }

    createSymbol();
}
```

**Comment by isiahmeadows** [comment]
@nzakas Have you thought of using a hash map of identifiers that you whitelist to ignore within the rule? You can add them when you see the `typeof`, and remove them after you leave the code path. Or can you not add code path-specific data that inherits prototypically or similarly as such?

**Comment by nzakas** [comment]
@isiahmeadows all of this is done inside of `escope`, which is mostly a black box for us. It doesn't have these dynamic capabilities. That's why I'm saying, what you're asking might be technically possible, but it would be a massive amount of work to do it correctly and I just can't see us taking on that amount of work at this point.

**Comment by isiahmeadows** [comment]
@nzakas Should this be opened as an issue against `escope` instead, to request an `isTypeofGuarded()` method?

**Comment by nzakas** [comment]
@isiahmeadows no, that's beyond on the scope of escope. If anything, escope would have to be changed to allow dynamic scope changes, however, there's very little work being done on escope, so I doubt any significant change that isn't completely necessary would happen. 

As I've said, the amount of work necessary to implement what you're requesting makes it unlikely to happen unless you want to take the effort on yourself.

**Comment by isiahmeadows** [comment]
Okay. Thanks!

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by medikoo** [comment]
I was just having this issue, and it's sad it was closed without a solution.

Using `/*global .. */` remains a  dirty workaround to this valid use case.
